lib/SPIRV/SPIRVWriter.cpp

```diff
@@ -3490,6 +3518,26 @@ static bool allowsApproxFunction(IntrinsicInst *II) {
            cast<VectorType>(Ty)->getElementType()->isFloatTy()));
 }
 
+bool allowDecorateWithBufferLocationINTEL(IntrinsicInst *II) {
+  SmallVector<Value *, 8> UserList;
+
+  for (auto *Inst : II->users()) {
+    // if castInst, push Sussessors
+    if (auto *Cast = dyn_cast<CastInst>(Inst)) {
+      for (auto *Sussessor : Cast->users())
+        UserList.push_back(Sussessor);
+    } else {
+      UserList.push_back(Inst);
+    }
+  }
+
+  for (auto &Inst : UserList)
+    if (isa<LoadInst>(Inst) || isa<StoreInst>(Inst))
+      return true;
```

Comment:
Lets imagine a situation, when we have a pointer to a composite object. The memory this pointer points to has a specific memory location and we annotate it appropriately. Now we access this composite via GEP. I would assume, that the result pointer should point to the same specific memory and hence should be decorated, but it's seems like it won't happen. Is it intentional or it wasn't not considered?

Reply:
The PR title says "decorating a scalar pointer", so maybe this case is not covered intentionally.

Reply:
The title is a bit misleading. Let me update it.

I think the confusion is coming from the title that we are trying to decorate scalar pointers. If we want to do this, Like Dimitry said, we will want to find all alias pointers and apply the Bufferlocation ID on all of them. This is the previous approach we tried, such that by applying Bufferlocation ID on a kernel arg and trying to find all alias pointer of it. But it seems we find a lot of difficulties in this implementation.

This new implementation is to decorate the pointer argument of OpLoad and Opstore with BufferLocationINTEL that given in its predecessor llvm.ptr.annotation intrinsic. The form of the IR is secured by a header file and builtin functions so we can keep the llvm.ptr.annotation to Opload/Opstore as close as possible in all occurrences.

Reply:
Okay, implementation-wise it makes sense. Should we change the description in the spec from "Indicates that the pointer must only point into ..." to "Indicates that the pointer and all of its aliases must only point into ..."? I'll duplicate this question in https://github.com/KhronosGroup/SPIRV-Registry/pull/185

Reply:
I double-checked with co-workers and spec. We don't need to handle GEP cases. Just need the BufferLocation ID on the annotated pointer. Also added comments in the spec PR

Reply:
Resolving this comment. Only nits remain to fix.